bgpd/bgp_evpn_mh.c

```diff
@@ -1533,6 +1533,9 @@ void bgp_evpn_path_es_link(struct bgp_path_info *pi, vni_t vni, esi_t *esi)
 	if (es_info->es == es)
 		return;
 
+	/* Set change flag for ES value is updated */
+	SET_FLAG(pi->flags, BGP_PATH_ATTR_CHANGED);
```

Comment:
1. bgp_evpn_path_es_link is use to link paths to an ES for easy per-ES processing. So this is not the right place to look for path_attr changes.
2. What you are pointing out (ES1 to ES2 dest change) is basic mac-mobility and addressd via the the es change check done in install_evpn_route_entry -
```
static int install_evpn_route_entry(struct bgp *bgp, struct bgpevpn *vpn,
                    const struct prefix_evpn *p,
                    struct bgp_path_info *parent_pi)
{
>>>>>>>>>>>>>>>>>>>>>>> SNIP >>>>>>>>>>>>>>>>>>>>>>
 if (!pi) {
        /* Create an info */
        pi = bgp_create_evpn_bgp_path_info(parent_pi, dest,
                            parent_pi->attr);
        new_local_es = bgp_evpn_attr_is_local_es(pi->attr);
    } else {
        if (attrhash_cmp(pi->attr, parent_pi->attr)
            && !CHECK_FLAG(pi->flags, BGP_PATH_REMOVED)) {
            bgp_dest_unlock_node(dest);
            return 0;
        }
        /* The attribute has changed. */
        /* Add (or update) attribute to hash. */
        attr_new = bgp_attr_intern(parent_pi->attr);

        /* Restore route, if needed. */
        if (CHECK_FLAG(pi->flags, BGP_PATH_REMOVED))
            bgp_path_info_restore(dest, pi);

        /* Mark if nexthop has changed. */
        if (!IPV4_ADDR_SAME(&pi->attr->nexthop, &attr_new->nexthop))
            SET_FLAG(pi->flags, BGP_PATH_IGP_CHANGED);

        old_local_es = bgp_evpn_attr_is_local_es(pi->attr);
        new_local_es = bgp_evpn_attr_is_local_es(attr_new);
        /* If ESI is different or if its type has changed we
         * need to reinstall the path in zebra
         */
        if ((old_local_es != new_local_es) 
            || memcmp(&pi->attr->esi, &attr_new->esi,
                  sizeof(attr_new->esi))) {

            if (BGP_DEBUG(evpn_mh, EVPN_MH_RT))
                zlog_debug("VNI %d path %pFX chg to %s es",
                       vpn->vni, &pi->net->p,
                       new_local_es ? "local"
                            : "non-local");
            bgp_path_info_set_flag(dest, pi, BGP_PATH_ATTR_CHANGED); >>>>>>>>>>>>>
        }

        /* Unintern existing, set to new. */
        bgp_attr_unintern(&pi->attr);
        pi->attr = attr_new;
        pi->uptime = bgp_clock();
    }
>>>>>>>>>>>>>>>>>>>>>>> SNIP >>>>>>>>>>>>>>>>>>>>>>
}
```

I have tested dest changes (ES1=>ES2, ES=>non-ES and non-ES=>ES) and don't see the issue that you mention. Can you please provide details on the problem you are seeing?

Reply:
Thanks for the information, it would be my faulty report, my local code base is not latest, and might miss some related commits.